### PR TITLE
[Model Validation] Add Model Validation to mlflow.evaluate

### PR DESCRIPTION
**Pull request by zhe-db**
## What changes are proposed in this pull request?
Add model validation functionality to `mlflow.evaluate`
- Add `MetricThreshold` class to let users define validation thresholds for model validation.
- Add `_validate` function for metric value threshold check and model comparison check
- Add internal class `_MetricValidationResult` per metric for formatting failure message.
- Add baseline_model and validation_thresholds argument to MLflow.evaluate API.
- Add evaluation of baseline models during model validation.

If the model validation passed, no exception will be thrown and `EvaluationResult` will be returned. If the model validation failed, `mlflowModelValidationFailedException` will be thrown containing detailed failure message indicating why the model validation did not pass.

Documentations:

Screen Shot 2022-08-23 at 5 26 01 PM
Screen Shot 2022-08-23 at 5 28 05 PM
Screen Shot 2022-08-23 at 5 28 14 PM

(Please fill in changes proposed in this fix)

## How is this patch tested?

- [ ] I have written tests (not required for typo or doc fix) and confirmed the proposed feature/bug-fix/change works.

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Click the `Details` link on the `Preview docs` check.
2. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by dbczumar** [comment]
@jerrylian-db Can you resolve conflicts with `master`?

**Comment by dbczumar** [comment]
@jerrylian-db Can we file and prioritize an internal ticket to add some docs for model validation to https://mlflow.org/docs/latest/models.html#model-evaluation? Right now, the only available docs are API docs, and we can't release until feature overview docs are added.

File: mlflow/models/evaluation/base.py [hunk 1]
```diff
@@ -896,8 +1074,59 @@ def scatter_plot(eval_df, builtin_metrics, artifacts_dir):
...
+                                                         custom_metrics=[custom_l1_loss],
+                                                         validation_thresholds=thresholds,
+                                                         baseline_model=your_baseline_model
+
+                                                     )
+
+    :param baseline_model: (Optional) A string URI referring to a MLflow model with the pyfunc
+                                      flavor as a baseline model to be compared with the
+                                      candidate model (specified by the `model` param) for model
+                                      validation. (pyfunc model instance is not allowed)
```
Comment (dbczumar) [inline mlflow/models/evaluation/base.py:1126]:
```suggestion
                                      validation.
```

Comment (dbczumar) [inline mlflow/models/evaluation/base.py:1126]:
```suggestion
    :param baseline_model: (Optional) A string URI referring to an MLflow model with the pyfunc
                                      flavor. If specified, the candidate ``model`` is compared to this baseline
                                      for model validation purposes.
```

File: mlflow/models/evaluation/base.py [hunk 2]
```diff
@@ -896,8 +1074,59 @@ def scatter_plot(eval_df, builtin_metrics, artifacts_dir):
                                        custom_metrics=[squared_diff_plus_one, scatter_plot],
                                    )
 
+    :param validation_thresholds: (Optional) A dictionary of metric name to
+                                             :py:class:`mlflow.models.MetricThreshold` used for
+                                             model validation. Metric name must be one of the
+                                             builtin metric's name or
+                                             name of a custom metric in custom_metrics argument.
```
Comment (dbczumar) [inline mlflow/models/evaluation/base.py:1081]:
```suggestion
                                             model validation. Each metric name must either be the name of
                                             a builtin metric or the name of a custom metric defined in the
                                             ``custom_metrics`` parameter.
```

File: mlflow/models/evaluation/base.py [hunk 3]
```diff
@@ -758,7 +934,9 @@ def evaluate(
         - The evaluation dataset label values must be numeric or boolean, all feature values
           must be numeric, and each feature column must only contain scalar values.
 
-    :param model: A pyfunc model instance, or a URI referring to such a model.
+    :param model: A pyfunc model instance, or a URI referring to such a model, as
+                  the candidate_model to be compared with baseline_model
+                  (specified by the `baseline_model` param) for model validation.
```
Comment (dbczumar) [inline mlflow/models/evaluation/base.py:939]: I think that the documentation of ``baseline_model`` is sufficient to explain the role of candidate / baseline models.

```suggestion
    :param model: A pyfunc model instance, or a URI referring to such a model.
```

File: mlflow/models/evaluation/base.py [hunk 4]
```diff
@@ -673,6 +847,8 @@ def evaluate(
     evaluators=None,
     evaluator_config=None,
     custom_metrics=None,
+    validation_thresholds=None,
+    baseline_model=None,
 ):
     """
```
Comment (dbczumar) [inline mlflow/models/evaluation/base.py:853]: @jerrylian-db Can we add some documentation to `mlflow.evaluate()` (outside of the params section) explaining how this API can be used for model validation?

**Comment by dbczumar** [comment]
@jerrylian-db Can we file and prioritize an internal ticket to add an example for model validation to https://github.com/mlflow/mlflow/tree/master/examples/evaluation? Right now, we don't have any examples. We can't release this feature until examples are added.

File: mlflow/models/evaluation/base.py [hunk 5]
```diff
@@ -906,9 +1135,25 @@ def scatter_plot(eval_df, builtin_metrics, artifacts_dir):
...
+        raise MlflowException(
+            message="The baseline model argument must be a string URI referring to an MLflow model",
+            error_code=INVALID_PARAMETER_VALUE,
+        )
+    elif _model_validation_contains_model_comparison(validation_thresholds):
+        raise MlflowException(
+            message="The baseline model argument must be a string URI referring to "
+            "an MLflow model when model comparison thresholds  "
+            "(min_absolute_change, min_relative_change) are specified.",
+            error_code=INVALID_PARAMETER_VALUE,
```
Comment (dbczumar) [inline mlflow/models/evaluation/base.py:1156]: When I ran `mlflow.evaluate()` and specified validation thresholds without specifying a baseline model, the error message was:

```
mlflow.exceptions.MlflowException: The baseline model argument must be a string URI referring to an MLflow model when model comparison thresholds  (min_absolute_change, min_relative_change) are specified
```

This is kind of confusing, as it makes it seem like the issue is a type error when, in fact, the issue is that the baseline was never specified. Can we distinguish between these cases (wrong type vs absent) and raise different errors?

File: mlflow/models/evaluation/base.py [hunk 6]
```diff
@@ -598,6 +661,107 @@ def _get_last_failed_evaluator():
...
+            # - if (metric_value - baseline) / baseline < min_relative_change for higher is better
+            # - if (baseline - metric_value) / baseline < min_relative_change for lower is better
+            if metric_threshold.higher_is_better:
+                relative_change = (
+                    candidate_metric_value - baseline_metric_value
+                ) / baseline_metric_value
+            else:
+                relative_change = (
+                    baseline_metric_value - candidate_metric_value
+                ) / baseline_metric_value
```
Comment (dbczumar) [inline mlflow/models/evaluation/base.py:747]: These computations fail when `baseline_metric_value` is zero, which is the case in the following overfitting example:

```
import mlflow
from mlflow.models.evaluation import EvaluationDataset
from mlflow.models import MetricThreshold

from sklearn import datasets
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split

# prepare train and test data
iris = datasets.load_iris()
X = iris.data
y = iris.target
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

import pandas as pd
xy_test = pd.DataFrame(X_test, columns=["0", "1", "2", "3"])
xy_test["label"] = y_test

thresholds = {
    # Metric Value Threshold
    "root_mean_squared_error": MetricThreshold(
        threshold=0.05,
        higher_is_better=False,
        min_relative_change=0.1
    ),
}

with mlflow.start_run():
    rid = mlflow.active_run().info.run_id

    logistic = LogisticRegression(C=1.0)
    logistic.fit(X_train, y_train)
    mod = mlflow.sklearn.log_model(logistic, artifact_path="mymodel")
    print(mod.model_uri)

    results = mlflow.evaluate(
        model="runs:/%s/mymodel" % rid,
        model_type="regressor",
        data=xy_test,
        evaluator_config={
            "log_model_explainability": False,
        },
        evaluators="default",
        validation_thresholds=thresholds,
        targets="label",
        baseline_model="runs:/acbea54588ee4122add50991d7094efd/mymodel",
    )
    print("RES", type(results))
    print(results.metrics)
    print(results.baseline_model_metrics)
    print(results.artifacts)
```

Comment (dbczumar) [reply mlflow/models/evaluation/base.py:747]: Traceback (most recent call last):
  File "eval.py", line 36, in <module>
    results = mlflow.evaluate(
  File "/Users/corey.zumar/mlflow/mlflow/models/evaluation/base.py", line 1189, in evaluate
    _validate(
  File "/Users/corey.zumar/mlflow/mlflow/models/evaluation/base.py", line 747, in _validate
    relative_change = (
ZeroDivisionError: float division by zero

**Comment by zhe-db** [comment]
> @jerrylian-db Can we file and prioritize an internal ticket to add an example for model validation to https://github.com/mlflow/mlflow/tree/master/examples/evaluation? Right now, we don't have any examples. We can't release this feature until examples are added.

Thanks for the review Corey. I still have time to work on this (the school starts on 29th); I will add documentations with examples for model validation to https://github.com/mlflow/mlflow/tree/master/examples/evaluation

File: mlflow/models/evaluation/base.py (hunk 6 above)
Comment (jerrylian-db) [reply mlflow/models/evaluation/base.py:759]: I've adapted your feedback. Here's the output of the overfitting example.
Screen Shot 2022-08-23 at 12 17 39 PM

**Comment by jerrylian-db** [comment]
@dbczumar Here are the changes I've made after taking in your feedback:
* Created a missing baseline model failure message and corresponding tests.
* Built a way to handle the case when baseline metric values equals 0 and corresponding tests.
* Added `MetricThreshold` parameter checks and corresponding tests.
* Updated doc strings for the `MetricThreshold` class. Verified that the docs are generated correctly. (Updated PR description).
Would you mind taking another look? Thinking that we merge this code for now and in future PRs (that I will create this week) add more documentation and examples and further polish the feature.

**Comment by jerrylian-db** [comment]
Actually, closing this PR and reopening it for ownership purposes.

**Pull request closed without merging**